packages/web/src/pages/settings-page/components/desktop/DeveloperApps/EditAppPage.tsx

```diff
@@ -0,0 +1,211 @@
+import { Divider } from 'components/divider'
+import Toast from 'components/toast/Toast'
+import { MountPlacement } from 'components/types'
+import { copyToClipboard } from 'utils/clipboardUtil'
+import styles from './EditAppPage.module.css'
+import { CreateAppPageProps, CreateAppsPages } from './types'
+import { Name, Status } from '@audius/common/models'
+import { accountSelectors } from '@audius/common/store'
+import { Form, Formik, useField } from 'formik'
+import { z } from 'zod'
+import { toFormikValidationSchema } from 'zod-formik-adapter'
+
+import { make, useRecord } from 'common/store/analytics/actions'
+import { TextAreaField, TextField } from 'components/form-fields'
+import { useSelector } from 'utils/reducer'
+import { DEVELOPER_APP_DESCRIPTION_MAX_LENGTH, DEVELOPER_APP_IMAGE_URL_MAX_LENGTH, DEVELOPER_APP_NAME_MAX_LENGTH, developerAppEditSchema, useEditDeveloperApp } from '@audius/common/api'
+
+const { getUserId } = accountSelectors
+
+type EditAppPageProps = CreateAppPageProps
+
+type DeveloperAppValues = z.input<typeof developerAppEditSchema>
+
+const messages = {
+  appNameLabel: 'App Name',
+  imageUrlLabel: 'App Icon URL',
+  apiKey: 'api key',
+  copyApiKeyLabel: 'copy api key',
+  copied: 'Copied!',
+  goBack: 'Back to Your Apps',
+  back: 'Back',
+  save: 'Save Changes',
+  saving: 'Saving',
+  miscError: 'Sorry, something went wrong. Please try again later.'
+}
+
+const ImageField = ({ name }: { name: string }) => {
+  const [{ value }] = useField(name)
+  return value
+    ? <img src={value} height='136' width='136' />
+    : <Flex
+        w='100%'
+        justifyContent='center'
+        alignItems='center'
+        borderRadius='l'
+        css={{ backgroundColor: 'var(--harmony-n-200)' }}
+      >
+        <IconEmbed
+          color='subdued'
+          css={{ width: '75px', height: '75px' }}
+        />
+      </Flex>
+}
+
+export const EditAppPage = (props: EditAppPageProps) => {
+  const { params, setPage } = props
+  const userId = useSelector(getUserId) as number
+  const { name, description, apiKey, imageUrl } = params || {}
+
+  const record = useRecord()
+
+  const [editDeveloperApp, result] = useEditDeveloperApp()
+  const [submitError, setSubmitError] = useState<string | null>(null)
+
+  const { status, data, errorMessage } = result
+
+  useEffect(() => {
+    if (status === Status.SUCCESS) {
+      setPage(CreateAppsPages.YOUR_APPS)
+      record(
+        make(Name.DEVELOPER_APP_EDIT_SUCCESS, {
+          name: name || '',
+          apiKey: apiKey || ''
+        })
+      )
+    }
+  }, [data, record, setPage, status])
+
+  useEffect(() => {
+    if (status === Status.ERROR) {
+      setSubmitError(messages.miscError)
+      record(
+        make(Name.DEVELOPER_APP_EDIT_ERROR, {
+          error: errorMessage
+        })
+      )
+    }
+  }, [errorMessage, record, status])
+
+  const handleSubmit = useCallback(
+    (values: DeveloperAppValues) => {
+      setSubmitError(null)
```

Comment:
I had thought formik had a way to manage form-level errors for us such that we don't have to implement this kind of tracking for every mutation we want to do.

Reply:
It absolutely does. Not sure why none of the components in these developer apps do it, but I'll make a ticket to do a bigger & better refactor for it